crates/fuel-core/src/state/historical_rocksdb.rs

```diff
@@ -439,15 +703,50 @@ where
         height: Option<Description::Height>,
         changes: Changes,
     ) -> StorageResult<()> {
-        let mut storage_transaction =
-            StorageTransaction::transaction(&self.db, ConflictPolicy::Overwrite, changes);
+        // cumulative_changes_lock_guard is defined to be Some only when the migration is in progress.
+        // If the migration is not in progress, the default set of changes will be used, and the overhead
+        // for handling caused by this function to handle the migration will be minimal.
+        let (cumulative_changes, maybe_cumulative_changes_lock_guard) =
+            self.take_migration_changes();
+
+        let mut migration_transaction = StorageTransaction::transaction(
+            &self.db,
+            ConflictPolicy::Overwrite,
+            // We need to keep ownership of the migration changes in case the current transaction fails.
+            // In this case we need to place the cumulative changes for migrating the modification history
+            // back into the lock guard.
+            // Note: this might be requiring cloning a fair amount of data
+            cumulative_changes.clone(),
+        );
+        let mut storage_transaction = StorageTransaction::transaction(
+            &mut migration_transaction,
+            ConflictPolicy::Overwrite,
+            changes,
+        );
+
         if let Some(height) = height {
             self.store_modifications_history(&mut storage_transaction, &height)?;
```

Comment:
In case we fail here, could it be that we lose the changes?

Reply:
Good catch. 2739d6162aa9e7e3fbc6e4c56cb546379f6f3146